Kuzzle/Offline/Query/QueryReplayer.cs

```diff
@@ -0,0 +1,228 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+using KuzzleSdk.API.Offline;
+using Newtonsoft.Json.Linq;
+
+namespace KuzzleSdk {
+
+  public interface IQueryReplayer {
+    int Count { get; }
+    bool Lock { get; set; }
+    bool WaitLoginToReplay { get; set; }
+
+    bool Enqueue(JObject query);
+    JObject Dequeue();
+
+    void RejectAllQueries(Exception exception);
+    void RejectQueries(Predicate<JObject> predicate, Exception exception);
+    int Remove(Predicate<JObject> predicate);
+
+    CancellationTokenSource ReplayQueries(bool resetWaitLogin = true);
+    CancellationTokenSource ReplayQueries(Predicate<JObject> predicate, bool resetWaitLogin = true);
+
+    void CancelReplay();
+  }
+
+  internal sealed class QueryReplayer : IQueryReplayer {
+
+    private Int64 startTime;
+    private IKuzzle kuzzle;
+    private List<TimedQuery> queue;
+    private IOfflineManager offlineManager;
+    private CancellationTokenSource cancellationTokenSource;
+    private bool currentlyReplaying = false;
+
+    /// <summary>
+    /// Does the QueryReplayer accepts new queries.
+    /// Should only be true if there is an 'auth:login' or 'auth:logout' call in the queue
+    /// </summary>
+    public bool Lock { get; set; } = false;
+
+    /// <summary>
+    /// True if the QueryReplayer has to wait a login call before replaying the queue
+    /// </summary>
+    public bool WaitLoginToReplay { get; set; } = false;
+
+    /// <summary>
+    /// Constructor of the QueryReplayer class.
+    /// </summary>
+    internal QueryReplayer(IOfflineManager offlineManager, IKuzzle kuzzle) {
+      queue = new List<TimedQuery>();
+      this.offlineManager = offlineManager;
+      this.kuzzle = kuzzle;
+      this.cancellationTokenSource = new CancellationTokenSource();
+      this.ReplayQuery = ReplayOneQuery;
+    }
+
+    /// <summary>
+    /// Add a new query to the queue.
+    /// Return true if success
+    /// </summary>
+    public bool Enqueue(JObject query) {
+      if (Lock) return false;
+
+      lock (queue) {
+        if ((queue.Count < offlineManager.MaxQueueSize) || offlineManager.MaxQueueSize < 0) {
```

Comment:
(nitpicking) needless parens

```suggestion
        if (queue.Count < offlineManager.MaxQueueSize || offlineManager.MaxQueueSize < 0) {
```